python/delta-kernel-rust-sharing-wrapper/src/lib.rs

```diff
@@ -78,6 +80,38 @@ impl ScanBuilder {
     }
 }
 
+fn try_get_schema(schema: &Arc<StructType>) -> Result<ArrowSchemaRef, delta_kernel::Error> {
+    Ok(Arc::new(schema.as_ref().try_into().map_err(|e| {
+        delta_kernel::Error::Generic(format!("Could not get result schema: {e}"))
+    })?))
+}
+
+// TODO(patrick.jin): Change return type to RecordBatchIterator<impl Iterator<...>>
+fn try_create_record_batch_iter(
+    results: impl Iterator<Item = DeltaResult<ScanResult>>,
+    result_schema: ArrowSchemaRef,
+) -> RecordBatchIterator<Vec<Result<RecordBatch, ArrowError>>> {
+    let record_batches: Vec<_> = results
+        .map(|res| {
+            let scan_res = res.and_then(|res| Ok((res.full_mask(), res.raw_data?)));
+            let (mask, data) =
+                scan_res.map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
+            let record_batch: RecordBatch = data
+                .into_any()
+                .downcast::<ArrowEngineData>()
+                .map_err(|_| ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string()))?
+                .into();
+            if let Some(mask) = mask {
+                let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
+                Ok(filtered_batch)
+            } else {
+                Ok(record_batch)
+            }
+        })
+        .collect();
+    RecordBatchIterator::new(record_batches, result_schema)
+}
+
```

Comment:
```suggestion
fn try_create_record_batch_iter(
    results: impl Iterator<Item = DeltaResult<ScanResult>>,
    result_schema: ArrowSchemaRef,
) -> RecordBatchIterator<impl Iterator<Item = Result<RecordBatch, ArrowError>>> {
    let record_batches = results.map(|res| {
        let scan_res = res.and_then(|res| Ok((res.full_mask(), res.raw_data?)));
        let (mask, data) = scan_res.map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
        let record_batch: RecordBatch = data
            .into_any()
            .downcast::<ArrowEngineData>()
            .map_err(|_| ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string()))?
            .into();
        if let Some(mask) = mask {
            let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
            Ok(filtered_batch)
        } else {
            Ok(record_batch)
        }
    });
    RecordBatchIterator::new(record_batches, result_schema)
}
```

Reply:
@OussamaSaoudi-db since this is a pretty common pattern I wonder if somewhere in default enging/arrow implementation we should basically provide this

Reply:
Yeah, let's create an issue for it. we have it in like 3 tests and our examples, and here, so clearly it would be useful to have a canonical implementation

Reply:
https://github.com/delta-io/delta-kernel-rs/issues/592